Arm/BitVec.lean

```diff
@@ -308,37 +308,35 @@ def flattenTR {n : Nat} (xs : List (BitVec n)) (i : Nat)
 /-- Reverse bits of a bit-vector. -/
 def reverse (x : BitVec n) : BitVec n :=
   let rec reverseTR (x : BitVec n) (i : Nat) (acc : BitVec n) :=
-    if i < n then
-      let xi := extractLsb i i x
-      have h : i - i + 1 = (n - i - 1) - (n - i - 1) + 1 := by omega
-      let acc := BitVec.partInstall (n - i - 1) (n - i - 1) (BitVec.cast h xi) acc
-      reverseTR x (i + 1) acc
-    else acc
-  reverseTR x 0 $ BitVec.zero n
-
-example : reverse 0b11101#5 = 0b10111#5 := by
-  -- (FIXME) With leanprover/lean4:nightly-2024-08-29, just `rfl` sufficed here.
-  simp [reverse, reverse.reverseTR]
-  rfl
+    match i with
+    | 0 => acc
+    | j + 1 =>
+      have h1 : i - 1 - (i - 1) + 1 = 1 := by omega
+      let xi : BitVec 1 := BitVec.cast h1 $ extractLsb (i - 1) (i - 1) x
+      have h2 : 1 = (n - i) - (n - i) + 1 := by omega
+      let acc := BitVec.partInstall (n - i) (n - i) (BitVec.cast h2 xi) acc
+      reverseTR x j acc
+  reverseTR x n $ BitVec.zero n
+
+example : reverse 0b11101#5 = 0b10111#5 := by rfl
 
 /-- Split a bit-vector into sub vectors of size e. -/
 def split (x : BitVec n) (e : Nat) (h : 0 < e): List (BitVec e) :=
   let rec splitTR (x : BitVec n) (e : Nat) (h : 0 < e)
     (i : Nat) (acc : List (BitVec e)) : List (BitVec e) :=
-    if i < n/e then
-      let lo := i * e
+    match i with
+    | 0 => acc
+    | j + 1 =>
+      let lo := (n / e - i) * e
       let hi := lo + e - 1
       have h₀ : hi - lo + 1 = e := by simp only [hi, lo]; omega
       let part : BitVec e := BitVec.cast h₀ (extractLsb hi lo x)
       let newacc := part :: acc
```

Comment:
(This is unrelated to the main change, so feel free to ignore)

Generally, using `extractLsb'` is nicer than `extractLsb`, because you specify the length directly and can thus generally avoid casts. 
```suggestion
      let part : BitVec e := extractLsb' lo e x
      let newacc := part :: acc
```

(do double check if the start should be `lo` or `lo +/- 1`, I always forget if there's any inclusive/exclusive interval shenanigans going on with these)

Reply:
Thanks for the suggestion. Several functions are simplified this way.